ng-dev/utils/yargs.ts

```diff
@@ -0,0 +1,39 @@
+import yargs, {Arguments, Argv} from 'yargs';
+import {Log} from './logging.js';
+
+// A function to be called when the command completes.
+type CompletedFn = (err: Error | null) => Promise<void> | void;
+
+/** List of functions to be called upon command completion. */
+const completedFunctions: CompletedFn[] = [];
+
+/** Register a function to be called when the command completes. */
+export function registerCompletedFunction(fn: CompletedFn) {
+  completedFunctions.push(fn);
+}
+
+/** Error to be thrown when yargs completes without running a command. */
+export class YargsError extends Error {
+  constructor() {
+    super('Error is parse or validation of command');
+  }
+}
+
+/**
+ * Run the yargs process, as configured by the supplied function, calling a set of completion
+ * functions after the command completes.
+ */
+export function runParserWithCompletedFunctions(applyConfiguration: (argv: Argv) => Argv) {
+  applyConfiguration(yargs([])).parse(
+    process.argv.slice(2),
+    async (err: Error | null, _: Arguments, output: string) => {
+      if (output) {
+        err = new YargsError();
+        Log.log(output);
```

Comment:
Sorry for being a little picky here, but Yargs is very imprecise on `.parse` as it seems. My thinking (correct me if I'm wrong) is that we should:

* Always log `output` if there is something (it doesn't need to be for an error technically)
* If there is an `err`, we print it? (not sure about that -- what did Yarn do by default?)
* If there is an `err`, we set the `process.exitCode` to `1` (because Yarn no longer exits on errors right?)

Reply:
I updated again, let me know how this looks to you.

Reply:
Looks close, subjectively. I would be interested in what you actually think is the right thing to do here. Only thing that looks suspicious too me right now is that we build an err if there is an output (does output always indicate an error?)

Reply:
Output always indicates that Yargs did not run the handler for a command, whether than be cause of a a `--help` flag or because of a validation error.

Since calling with `--help` is not an error, we are creating an error to represent that the command did not run.

Reply:
If that's the case, then this seems reasonable. From the documentation it doesn't seem to indicate that strictly the handler wasn't invoked though.

Don't want to overcomplicate this, feel free to merge.

Reply:
I read more up and it seems like the most idiomatic solution would really be to just use a top level async/await and try/catch/finally.

https://github.com/yargs/yargs/pull/1823

I would like the more obvious/predictable top level await more, but non blocking

Reply:
Okay, I moved to the top level await method, let me know what you think.